subsys/net/lib/nrf_provisioning/src/nrf_provisioning_coap.c

```diff
@@ -0,0 +1,643 @@
+}
+
+static int socket_connect(int *const fd)
+{
+	static struct addrinfo hints;
+	int st;
+	int ret = 0;
+	struct sockaddr *sa;
+	char peer_addr[INET6_ADDRSTRLEN];
+
+#if defined(CONFIG_NET_IPV6) && defined(CONFIG_NET_IPV4)
+	hints.ai_family = AF_UNSPEC;
+#elif defined(CONFIG_NET_IPV6)
+	hints.ai_family = AF_INET6;
+#elif defined(CONFIG_NET_IPV4)
+	hints.ai_family = AF_INET;
+#else
+	hints.ai_family = AF_UNSPEC;
+#endif /* defined(CONFIG_NET_IPV6) && defined(CONFIG_NET_IPV4) */
+
+	hints.ai_socktype = SOCK_DGRAM;
+	st = getaddrinfo(COAP_HOST, PEER_PORT, &hints, &address);
+	LOG_DBG("getaddrinfo status: %d", st);
+	LOG_DBG("fd: %d", *fd);
+
+	if (st != 0) {
+		LOG_ERR("Unable to resolve address");
+		ret = -EFAULT;
+		goto clean_up;
+	}
+
+	sa = address->ai_addr;
+	inet_ntop(sa->sa_family,
+		  (void *)&((struct sockaddr_in *)sa)->sin_addr,
+		  peer_addr,
+		  INET6_ADDRSTRLEN);
+	LOG_DBG("getaddrinfo() %s", peer_addr);
+
+	*fd = socket(address->ai_family, address->ai_socktype, IPPROTO_DTLS_1_2);
+	if (*fd < 0) {
+		LOG_ERR("Failed to create UDP socket %d", errno);
+		ret = -ENOTCONN;
+		goto clean_up;
+	}
+
+	/* Setup TLS socket options */
+	ret = tls_setup(*fd);
+	if (ret) {
+		LOG_ERR("Failed to setup TLS socket option");
+		ret = -EACCES;
+		goto clean_up;
+	}
+
+	if (connect(*fd, address->ai_addr, address->ai_addrlen) < 0) {
+		LOG_ERR("Failed to connect UDP socket %d", errno);
+		if (errno == ETIMEDOUT) {
+			ret = -ETIMEDOUT;
+		} else {
+			ret = -ECONNREFUSED;
+		}
+		goto clean_up;
+	}
+	LOG_INF("Connected");
+
+clean_up:
+
+	if (ret) {
+		if (*fd > -1) {
+			(void)close(*fd);
+			*fd = -1;
+		}
+	}
+
+	return ret;
+}
+
+static int socket_close(int *const fd)
+{
+	int ret;
+
+	if (*fd > -1) {
+		ret = close(*fd);
+		if (ret) {
+			LOG_WRN("Failed to close socket, error: %d", errno);
+		}
+		*fd = -1;
+	}
+
+	return 0;
+}
+
+static void coap_callback(int16_t code, size_t offset, const uint8_t *payload, size_t len,
+			  bool last_block, void *user_data)
+{
+	struct nrf_provisioning_coap_context *coap_ctx = NULL;
+
+	if (user_data) {
+		coap_ctx = (struct nrf_provisioning_coap_context *)user_data;
+	}
+	LOG_DBG("Callback code %d", code);
+	if (!coap_ctx) {
+		LOG_WRN("CoAP context not provided");
+		k_sem_give(&coap_response);
+		return;
+	}
+
+	coap_ctx->code = code;
+
+	if (code == COAP_RESPONSE_CODE_CONTENT || code == COAP_RESPONSE_CODE_CHANGED ||
+	    code == COAP_RESPONSE_CODE_CREATED) {
+		if (len) {
+			LOG_DBG("Response received, offset %d len %d", offset, len);
+			if (len > coap_ctx->rx_buf_len) {
+				LOG_ERR("RX buffer too small");
+				coap_ctx->code = -ENOMEM;
+				k_sem_give(&coap_response);
+				return;
+			}
+
+			memcpy(coap_ctx->rx_buf, payload, len);
+			coap_ctx->response = coap_ctx->rx_buf;
+			coap_ctx->response_len = len;
+		} else {
+			LOG_DBG("Operation succesfull");
+			coap_ctx->response_len = 0;
+		}
+		if (last_block) {
+			LOG_DBG("Last packet received");
+		}
+	}
+
+	k_sem_give(&coap_response);
```

Comment:
Do you expect that block transfers will occur during any part of provisioning, or will it always be within a single block?

You might want to only give the semaphore when last_block is true, if multiple blocks might occur. Otherwise, your send_coap_request() function will exit after the first block of a multiblock transfer, which might not be what you want. Further, the memcpy into the rx_buf would need to account for that.